code/implementation/host-stats/src/main/java/io/cattle/platform/host/stats/api/HostStatsLinkHandler.java

```diff
@@ -1,75 +1,99 @@
 
     HostApiService hostApiService;
     ObjectManager objectManager;
+    TokenService tokenService;
 
     @Override
     public String[] getTypes() {
-        return new String[] { InstanceConstants.TYPE_CONTAINER, HostConstants.TYPE };
+        return new String[] { HostConstants.TYPE, ProjectConstants.TYPE };
     }
 
     @Override
     public boolean handles(String type, String id, String link, ApiRequest request) {
-        return HostStatsConstants.LINK_STATS.equals(link);
+        return StatsConstants.HOST_STATS.equalsIgnoreCase(link) && (ProjectConstants.TYPE.equalsIgnoreCase(type) || HostConstants.TYPE.equalsIgnoreCase(type));
     }
 
     @Override
     public Object link(String name, Object obj, ApiRequest request) throws IOException {
-        Host host = null;
-        Instance instance = null;
-
-        if (obj instanceof Instance) {
-            instance = (Instance) obj;
-            host = DockerUtils.getHostFromContainer(objectManager, instance);
-        } else if (obj instanceof Host) {
-            host = (Host) obj;
+        List<Host> hosts = new ArrayList<>();
+        boolean project = false;
+        if (obj instanceof Host) {
+            Host origHost = (Host) obj;
+            hosts.add(origHost);
         }
 
-        if (host == null) {
-            return null;
+        if (obj instanceof Account && AccountConstants.PROJECT_KIND.equals(((Account) obj).getKind())) {
+            hosts.addAll(objectManager.mappedChildren(obj, Host.class));
+            project = true;
         }
 
-        String[] pathSegments = null;
-        if (instance != null) {
-            pathSegments = new String[] { HOST_STATS_PATH.get(), DockerUtils.getDockerIdentifier(instance) };
-        } else {
-            pathSegments = new String[] { HOST_STATS_PATH.get() };
+        if (hosts.size() == 0) {
+            return null;
         }
 
-        HostApiAccess apiAccess = hostApiService.getAccess(request, host.getId(), Collections.<String, Object> emptyMap(), pathSegments);
-        if (apiAccess == null) {
-            return null;
+        String metaUrl = null;
+        StatsAccess meta = new StatsAccess();
+        List<StatsAccess> serviceStatsQuery = new ArrayList<>();
+        for (Host host : hosts) {
+            if (!CommonStatesConstants.ACTIVE.equals(host.getState())) {
+                continue;
+            }
+            Map<String, Object> payload = new HashMap<>();
+            payload.put("resourceId", ApiContext.getContext().getIdFormatter().formatId(objectManager.getType(host), host.getId()));
+            HostApiAccess apiAccess = hostApiService.getAccess(request, host.getId(), payload, new String[] { StatsConstants.HOST_STATS_PATH.get() });
+            if (apiAccess == null) {
+                continue;
+            }
+
+            String url = apiAccess.getUrl();
+
+            metaUrl = url + PROJECT_PATH;
+
+            StatsAccess statsAccess = new StatsAccess();
```

Comment:
Do you think it would make sense to create a constructor in StatsAccess having token/url as input params? As this code seems to repeat several times:

 statsAccess.setToken(apiAccess.getAuthenticationToken());
statsAccess.setUrl(url.toString());